gbus/tx/mysql/timeout.go

```diff
@@ -0,0 +1,223 @@
+package mysql
+
+import (
+	"database/sql"
+	"regexp"
+	"strings"
+	"time"
+
+	"github.com/sirupsen/logrus"
+
+	"github.com/wework/grabbit/gbus"
+)
+
+var _ gbus.TimeoutManager = &TimeoutManager{}
+
+//TimeoutManager is a mysql implementation of a persistent timeoutmanager
+type TimeoutManager struct {
+	Bus               gbus.Bus
+	Log               func() logrus.FieldLogger
+	TimeoutSaga       func(*sql.Tx, string) error
+	Txp               gbus.TxProvider
+	SvcName           string
+	timeoutsTableName string
+	exit              chan bool
+}
+
+func (tm *TimeoutManager) ensureSchema() error {
+	tblName := tm.timeoutsTableName
+	tx, e := tm.Txp.New()
+	if e != nil {
+		tm.Log().WithError(e).Error("failed to create schema for mysql timeout manager")
+		return e
+	}
+
+	createTableSQL := `CREATE TABLE IF NOT EXISTS ` + tblName + ` (
+      rec_id INT PRIMARY KEY AUTO_INCREMENT,
+      saga_id VARCHAR(255) UNIQUE NOT NULL,
+	  timeout DATETIME NOT NULL,
+	  INDEX (timeout),
+	  INDEX (saga_id)
+	 )`
+
+	if _, e := tx.Exec(createTableSQL); e != nil {
+		if rbkErr := tx.Rollback(); rbkErr != nil {
+			tm.Log().Warn("timeout manager failed to rollback transaction")
+		}
+		return e
+	}
+	return tx.Commit()
+}
+
+func (tm *TimeoutManager) purge() error {
+	purgeSQL := `DELETE FROM ` + tm.timeoutsTableName
+
+	tx, e := tm.Txp.New()
+	if e != nil {
+		tm.Log().WithError(e).Error("failed to purge timeout manager")
+		return e
+	}
+
+	if _, execErr := tx.Exec(purgeSQL); execErr != nil {
+		tm.Log().WithError(execErr).Error("failed to purge timeout manager")
+		return tx.Rollback()
+	}
+	return tx.Commit()
+}
+
+//Start starts the timeout manager
+func (tm *TimeoutManager) Start() error {
+	go tm.trackTimeouts()
+	return nil
+}
+
+//Stop shuts down the timeout manager
+func (tm *TimeoutManager) Stop() error {
+	tm.exit <- true
+	return nil
+}
+
+func (tm *TimeoutManager) trackTimeouts() {
+	tick := time.NewTicker(time.Second * 1).C
+	for {
+		select {
+		case <-tick:
+			tx, txe := tm.Txp.New()
+			if txe != nil {
+				tm.Log().WithError(txe).Warn("timeout manager failed to create a transaction")
+				continue
+			}
+			now := time.Now().UTC()
+			getTimeoutsSQL := `select saga_id from ` + tm.timeoutsTableName + ` where timeout < ? LIMIT 100`
```

Comment:
This should be select for update skip locked since you can have multiple instances running and you want them to handle only the ones that you actually want to work on and not to process the same line multiple times

Reply:
@vladshub  the locking of an instance and preventing from execting twice is done elsewhere (in the executeTimeouts function) as each timeout runs in their own transaction placing the FOR UPDATE here will have no effect

Reply:
If you add `FOR UPDATE SKIP LOCKS` it will process only the rows that it "caught" according to the [documentation](https://dev.mysql.com/doc/refman/8.0/en/innodb-locking-reads.html).

Reply:
that is exactly the point I do not want the executeTimeouts to include the transactions as each timeout needs to be executed in its own transaction and if you include the FOR UPDATE here the executeTimeouts will deadlock.
The protection against double execution is in the executeTimeout function when the code tries to lock the record and if the lock fails it skips the execution.